src/nnet3/discriminative-training.cc

```diff
@@ -0,0 +1,602 @@
+// limitations under the License.
+
+#include "nnet3/discriminative-training.h"
+#include "lat/lattice-functions.h"
+#include "cudamatrix/cu-matrix.h"
+
+namespace kaldi {
+namespace discriminative {
+
+DiscriminativeTrainingStats::DiscriminativeTrainingStats() {
+  std::memset(this, 0, sizeof(*this));
+  config.accumulate_gradients = false; 
+  config.accumulate_output = false; 
+  config.num_pdfs = 0; 
+}
+
+DiscriminativeTrainingStats::DiscriminativeTrainingStats(int32 num_pdfs) {
+  config.accumulate_gradients = false; 
+  config.accumulate_output = false; 
+  config.num_pdfs = num_pdfs; 
+  gradients.Resize(num_pdfs); 
+  output.Resize(num_pdfs);
+}
+
+// Constructor from config structure
+DiscriminativeTrainingStats::DiscriminativeTrainingStats(
+    DiscriminativeTrainingStatsOptions opts) : config(opts) { 
+  gradients.Resize(opts.num_pdfs); 
+  output.Resize(opts.num_pdfs);
+}
+
+// Reset statistics
+void DiscriminativeTrainingStats::Reset() {
+  gradients.SetZero();
+  output.SetZero();
+
+  tot_t = 0.0;
+  tot_t_weighted = 0.0;
+  tot_objf = 0.0;
+  tot_num_count = 0.0;
+  tot_den_count = 0.0;
+  tot_num_objf = 0.0;
+}
+
+// Sets the config structure
+void DiscriminativeTrainingStats::SetConfig(
+    const DiscriminativeTrainingStatsOptions &opts) {
+  config = opts;
+  gradients.Resize(opts.num_pdfs); 
+  output.Resize(opts.num_pdfs);
+}
+
+// This class is responsible for the forward-backward of the
+// 'supervision' lattices and computation of the objective function
+// and gradients.
+//
+// note: the supervision.weight is ignored by this class, you have to apply
+// it externally.
+class DiscriminativeComputation {
+  typedef Lattice::Arc Arc;
+  typedef Arc::StateId StateId;
+ 
+ public:
+  // Initialize the objcect.  Note: we expect the 'nnet_output' to have the
+  // same number of rows as supervision.num_frames * supervision.num_sequences,
+  // and the same number of columns as tmodel.NumPdfs(); but the
+  // ordering of the rows of 'nnet_output' is not the same as the ordering of
+  // frames in paths in the 'supervision' object (which has all frames of the
+  // 1st sequence first, then the 2nd sequence, and so on).  Instead, the
+  // frames in 'nnet_output' are ordered as: first the first frame of each
+  // sequence, then the second frame of each sequence, and so on. 
+  // This is done to be similar to the setup in 'chain' training 
+  // even though this does not offer any computational advantages here
+  // as in the 'chain' case.
+  DiscriminativeComputation(const DiscriminativeTrainingOptions &opts,
+      const TransitionModel &tmodel,
+      const CuVectorBase<BaseFloat> &log_priors,
+      const DiscriminativeSupervision &supervision,
+      const CuMatrixBase<BaseFloat> &nnet_output,
+      DiscriminativeTrainingStats *stats,
+      BaseFloat *l2_term,
+      CuMatrixBase<BaseFloat> *nnet_output_deriv,
+      CuMatrixBase<BaseFloat> *xent_output_deriv);
```

Comment:
I meant the xent output is probably not well trained enough in the case of
chain model. So it may not be good to do sequence training on top of it

On Mon, Mar 7, 2016, 19:50 Daniel Povey <notifications@github.com> wrote:

> In src/nnet3/discriminative-training.cc
> <https://github.com/kaldi-asr/kaldi/pull/563#discussion_r55300406>:
>
> > +  // frames in paths in the 'supervision' object (which has all frames of the
> > +  // 1st sequence first, then the 2nd sequence, and so on).  Instead, the
> > +  // frames in 'nnet_output' are ordered as: first the first frame of each
> > +  // sequence, then the second frame of each sequence, and so on.
> > +  // This is done to be similar to the setup in 'chain' training
> > +  // even though this does not offer any computational advantages here
> > +  // as in the 'chain' case.
> > +  DiscriminativeComputation(const DiscriminativeTrainingOptions &opts,
> > +      const TransitionModel &tmodel,
> > +      const CuVectorBase<BaseFloat> &log_priors,
> > +      const DiscriminativeSupervision &supervision,
> > +      const CuMatrixBase<BaseFloat> &nnet_output,
> > +      DiscriminativeTrainingStats *stats,
> > +      BaseFloat *l2_term,
> > +      CuMatrixBase<BaseFloat> *nnet_output_deriv,
> > +      CuMatrixBase<BaseFloat> *xent_output_deriv);
>
> I'm not sure why you said what you just said about "I remember decoding
> with the softmax part of the output, but the results were 10% absolute
> worse than chain part. I will try again to make sure I did correctly." I
> think you may have misinterpreted something I said. This is not relevant.
> … <#msg-f:1528192811736950802_>
> On Mon, Mar 7, 2016 at 7:47 PM, Vimal Manohar <notifications@github.com>
> wrote: In src/nnet3/discriminative-training.cc <#563 (comment)
> <https://github.com/kaldi-asr/kaldi/pull/563#discussion_r55300096>>: > +
> // frames in paths in the 'supervision' object (which has all frames of the
> > + // 1st sequence first, then the 2nd sequence, and so on). Instead, the
> > + // frames in 'nnet_output' are ordered as: first the first frame of
> each > + // sequence, then the second frame of each sequence, and so on. >
> + // This is done to be similar to the setup in 'chain' training > + //
> even though this does not offer any computational advantages here > + // as
> in the 'chain' case. > + DiscriminativeComputation(const
> DiscriminativeTrainingOptions &opts, > + const TransitionModel &tmodel, > +
> const CuVectorBase<BaseFloat> &log_priors, > + const
> DiscriminativeSupervision &supervision, > + const CuMatrixBase<BaseFloat>
> &nnet_output, > + DiscriminativeTrainingStats *stats, > + BaseFloat
> *l2_term, > + CuMatrixBase<BaseFloat> *nnet_output_deriv, > +
> CuMatrixBase<BaseFloat> *xent_output_deriv); I remember decoding with the
> softmax part of the output, but the results were 10% absolute worse than
> chain part. I will try again to make sure I did correctly. — Reply to this
> email directly or view it on GitHub <
> https://github.com/kaldi-asr/kaldi/pull/563/files#r55300096>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/563/files#r55300406>.
>
-- 
Vimal Manohar
PhD Student
Electrical & Computer Engineering
Johns Hopkins University